asynciterator.ts

```diff
@@ -1852,17 +1894,166 @@ class HistoryReader<T> {
   }
 }
 
+/**
+ * The IteratorIterator class provides a performant wrapper
+ * to convert ES Iterable instances into AsyncIterator instances.
+ */
+export class IteratorIterator<T> extends AsyncIterator<T> {
+  private _source: Iterator<T>;
+
+  /**
+   *
+   * @param source
+   */
+  constructor(source: Iterator<T>) {
+    super();
+    this._source = source;
+    this.readable = true;
+  }
+
+  read(): T | null {
+    let value: T | null = null;
+    if (this._source) {
+      const item = this._source.next();
+      if (!item.done)
+        value = item.value;
+      else
+        this.close();
+    }
+    return value;
+  }
+
+  protected _end(destroy: boolean = false) {
+    super._end(destroy);
+    delete this._source;
+  }
+}
+
+/**
+ * The WrappingIterator class can be used to wrap all supported kinds of data
+ * sources, both promisified and not, into AsyncIterator instances.
+ * This class is not meant to be used directly but, rather, through the
+ * `wrap()` function as it provides the mechanism that allows the latter
+ * to synchronously return an instance of AsyncIterator even for promisified
+ * sources.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source?: AsyncIteratorLike<T>;
+
+  constructor(sourceOrPromise: WrapSource<T> | Promise<WrapSource<T>> | null | undefined, options: WrapOptions = {}) {
+    super();
+    this._onSourceEnd = this._onSourceEnd.bind(this);
+    this._onSourceError = this._onSourceError.bind(this);
+    this._onSourceReadable = this._onSourceReadable.bind(this);
```

Comment:
Avoid creating custom versions of these functions for each class - instead use globally defined functions like https://github.com/RubenVerborgh/AsyncIterator/blob/4652546f82cf6bbd949a61443a21ce57f40f12e8/asynciterator.ts#L850-L854.

In fact - I'd imagine that `iteratorIterator` and `MappingIterator` could share a common abstract class after you rebase.

Reply:
I'm not sure as the sources we're dealing with in `IteratorIterator`, `MappingIterator` and `WrappingIterator` are of different kinds, with the latter class being the one where the source is less strictly-defined. Leaving this open and unaddressed until I rebase after #59 is merged.